File: android/src/main/kotlin/in/breeze/blaze_sdk_flutter/BlazeSdkFlutterPlugin.kt
package `in`.breeze.blaze_sdk_flutter

import androidx.annotation.NonNull

import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result
import io.flutter.embedding.engine.plugins.activity.ActivityAware
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding
import `in`.breeze.blaze.Blaze
import org.json.JSONObject
import android.app.Activity
import android.webkit.WebView


/** BlazeSdkFlutterPlugin */
class BlazeSdkFlutterPlugin: FlutterPlugin, MethodCallHandler, ActivityAware {
  private lateinit var blaze: Blaze
  private var activity: Activity? = null
 
  private lateinit var channel : MethodChannel

  override fun onAttachedToEngine(flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
    channel = MethodChannel(flutterPluginBinding.binaryMessenger, "blaze_sdk_flutter")
    channel.setMethodCallHandler(this)
    WebView.setWebContentsDebuggingEnabled(true);
  }

  override fun onAttachedToActivity(binding: ActivityPluginBinding) {
    activity = binding.activity
  }

  override fun onDetachedFromActivity() {
    activity = null
  }

  override fun onDetachedFromActivityForConfigChanges() {
    activity = null
  }

  override fun onReattachedToActivityForConfigChanges(binding: ActivityPluginBinding) {
    activity = binding.activity
  }

  override fun onMethodCall(call: MethodCall, result: Result) {
    when (call.method) {
      "initiate" -> {
        initiate(call.arguments)
        result.success(true)
      }
      "process" -> {
        process(call.arguments)
        result.success(true)
      }
      "handleBackPress" -> {
        result.success(handleBackPress())
      }
      "terminate" -> {
        terminate()
        result.success(true)
      }
      else -> {
        result.notImplemented()
      }
    }
  }

  override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
    channel.setMethodCallHandler(null)
  }

  // Utils

  fun safeParseJson(jsonString: Any): JSONObject {
    try {
      if(jsonString is String) {
        return JSONObject(jsonString)
      }
    } catch (e: Exception) {
      println("BlazeSDK: Failure in parsing JSON: ${e}")
    }
    return JSONObject()
  }
  
  // Exposed Bridge Methods

  fun initiate(initiatePayload: Any) {
    blaze = Blaze()
    if(activity != null) {
      val initiatePayloadJson = safeParseJson(initiatePayload)
      blaze.initiate(activity!!, initiatePayloadJson) { result ->
        activity?.runOnUiThread {
          channel.invokeMethod("blaze-callback", result.toString())
        }
      }
    } else {
      println("BlazeSDK: Activity is null")
    }
  }

  fun process(processPayload: Any) {
    val processPayloadJson = safeParseJson(processPayload)
    blaze.process(processPayloadJson)
  }

  fun handleBackPress(): Boolean  {
    return blaze.handleBackPress()
  }

  fun terminate() {
    blaze.terminate();
  }

}

File: ios/Classes/BlazeSdkFlutterPlugin.swift
import BlazeSDK
import Flutter
import Foundation
import UIKit

public class BlazeSdkFlutterPlugin: NSObject, FlutterPlugin {

    private var blaze: Blaze?
    private var methodChannel: FlutterMethodChannel?

    public static func register(with registrar: FlutterPluginRegistrar) {
        let channel = FlutterMethodChannel(
            name: "blaze_sdk_flutter", binaryMessenger: registrar.messenger())
        let instance = BlazeSdkFlutterPlugin()
        instance.methodChannel = channel
        registrar.addMethodCallDelegate(instance, channel: channel)
    }

    public func handle(
        _ call: FlutterMethodCall, result: @escaping FlutterResult
    ) {
        switch call.method {
        case "initiate":
            initiate(initiatePayload: call.arguments)
            result(true)
        case "process":
            process(processPayload: call.arguments)
            result(true)
        case "handleBackPress":
            result(handleBackPress())
        case "terminate":
            terminate()
            result(true)
        default:
            result(FlutterMethodNotImplemented)
        }
    }

    private func initiate(initiatePayload: Any?) {
        blaze = Blaze()
        let initiatePayloadJson = safeParseJson(initiatePayload)
        DispatchQueue.main.async {
            if let blaze = self.blaze,
                let rootViewController = UIApplication.shared.delegate?.window??
                    .rootViewController
            {
                blaze.initiate(
                    context: rootViewController,
                    initiatePayload: initiatePayloadJson
                ) { result in
                    if let resultData = try? JSONSerialization.data(
                        withJSONObject: result, options: []),
                        let resultString = String(
                            data: resultData, encoding: .utf8)
                    {
                        self.methodChannel?.invokeMethod(
                            "blaze-callback", arguments: resultString)
                    }
                }
            }
        }
    }

    private func process(processPayload: Any?) {
        let processPayloadJson = safeParseJson(processPayload)
        blaze?.process(payload: processPayloadJson)
    }

    private func handleBackPress() -> Bool {
        return true
    }

    private func terminate() {
        blaze?.terminate()
    }

    // Utility method to parse JSON
    func safeParseJson(_ json: Any?) -> [String: Any] {
        guard let jsonString = json as? String else {
            return [:]
        }
        if let data = jsonString.data(using: .utf8) {
            do {
                if let json = try JSONSerialization.jsonObject(
                    with: data, options: [])
                    as? [String: Any]
                {
                    return json
                }
            } catch {
                print("JSON parsing error: \(error)")
            }
        }
        return [:]
    }

}
